src/crimson/osd/replicated_recovery_backend.cc

```diff
@@ -516,14 +484,57 @@ seastar::future<ObjectRecoveryProgress> ReplicatedRecoveryBackend::build_push_op
 		     pop->version, pop->data.length());
       return seastar::make_ready_future<ObjectRecoveryProgress>
 		(std::move(new_progress));
-    }, PGBackend::read_errorator::all_same_way([](auto e) {
-	logger().debug("build_push_op: read exception");
-	return seastar::make_exception_future<ObjectRecoveryProgress>(e);
-      })
-    );
+    });
   });
 }
 
+seastar::future<uint64_t>
+ReplicatedRecoveryBackend::read_object_for_push_op(
+    const hobject_t& oid,
+    const interval_set<uint64_t>& copy_subset,
+    uint64_t offset,
+    uint64_t max_len,
+    PushOp* push_op)
+{
+  if (max_len == 0 || copy_subset.empty()) {
+    push_op->data_included.clear();
+    return seastar::make_ready_future<uint64_t>(offset);
+  }
+  // 1. get the extents in the interested range
+  return backend->fiemap(coll, ghobject_t{oid},
+                         0, copy_subset.range_end()).then_wrapped(
+    [=,extents=copy_subset](auto&& f) mutable {
+    try {
+      interval_set<uint64_t> fiemap_included(f.get0());
+      extents.intersection_of(fiemap_included);
+    } catch (std::exception &) {
```

Comment:
Do we expect an exception solely from `seastar::future::get0()` (which boils down to [`future_state::take()`](https://github.com/ceph/seastar/blob/master/include/seastar/core/future.hh#L568)) or from `intersection_of()` as well?

`CyanStore::fiemap()` throws `std::runtime_error` only:

```cpp
seastar::future<std::map<uint64_t, uint64_t>>
CyanStore::fiemap(
    CollectionRef ch,
    const ghobject_t& oid,
    uint64_t off,
    uint64_t len)
{
  auto c = static_cast<Collection*>(ch.get());

  ObjectRef o = c->get_object(oid);
  if (!o) {
    throw std::runtime_error(fmt::format("object does not exist: {}", oid));
  }
```

Maybe it's the time to errorate `fiemap()`? If so, then rather in a separated PR. Anyway, this is not a blocker.

Reply:
> Do we expect an exception solely from `seastar::future::get0()` (which boils down to [`future_state::take()`](https://github.com/ceph/seastar/blob/master/include/seastar/core/future.hh#L568)) or from `intersection_of()` as well?
> 

just from the former. to be specific, the `fiemap()` call.

> Maybe it's the time to errorate `fiemap()`? If so, then rather in a separated PR. Anyway, this is not a blocker.

agreed. we should not catch **all** exceptions. i will come up with a follow-up PR to errorate `fiemap()`.

Reply:
the reason i put `extents.intersection_of(fiemap_included);` in the `try` block is just to limit the lexical scope of `fiemap_included`. as what is interesting is `extents` here. i revised the comment and the `try` block a little bit. hopefully it's less confusing now.